File: sdk/core/llm/formatters/gemini-native.ts
/**
 * Gemini Native Format Converter
 *
 * Implements the conversion of request and response formats for the Gemini Native API
 * Uses Gemini's native endpoints
 */

import { BaseFormatter } from "./base.js";
import type { LLMRequest, LLMResult, LLMMessage, LLMToolCall } from "@wf-agent/types";
import type { ToolSchema } from "@wf-agent/types";
import type { FormatterConfig, BuildRequestResult, ParseStreamChunkResult } from "./types.js";
import { generateId } from "@wf-agent/common-utils";
import { convertToolsToGeminiFormat } from "../utils/index.js";
import { extractAndFilterSystemMessages } from "../message-helper.js";

/**
 * Gemini Native Format Converter
 */
export class GeminiNativeFormatter extends BaseFormatter {
  getSupportedProvider(): string {
    return "GEMINI_NATIVE";
  }

  buildRequest(request: LLMRequest, config: FormatterConfig): BuildRequestResult {
    const body = this.buildRequestBody(request, config);

    // Construct endpoint paths
    const method = config.stream ? "streamGenerateContent" : "generateContent";
    const endpoint = `/models/${config.profile.model}:${method}`;

    // The Gemini Native API uses query parameters to pass the API Key.
    const queryParams: Record<string, string | number | boolean> = {
      key: config.profile.apiKey || "",
    };

    // Streaming requests require the addition of the `alt=sse` parameter to obtain a response in SSE format.
    if (config.stream) {
      queryParams["alt"] = "sse";
    }

    // Merge user-defined query parameters
    if (config.queryParams) {
      Object.assign(queryParams, config.queryParams);
    }

    // Constructing request headers
    const headers: Record<string, string> = {
      "Content-Type": "application/json",
      ...config.profile.headers,
    };

    // Add authentication headers (supporting both x-goog-api-key and Authorization Bearer methods).
    if (config.profile.apiKey) {
      Object.assign(headers, this.buildAuthHeader(config.profile.apiKey, config, "x-goog-api-key"));
    }

    // Add custom request headers
    Object.assign(headers, this.buildCustomHeaders(config));

    // Apply a custom request body
    const finalBody = this.applyCustomBody(body, config);

    return {
      httpRequest: {
        url: endpoint,
        method: "POST",
        headers,
        query: queryParams,
        body: finalBody,
        timeout: config.timeout,
      },
      transformedBody: finalBody,
    };
  }

  parseResponse(data: unknown, config: FormatterConfig): LLMResult {
    const dataRecord = data as Record<string, unknown>;
    const candidates = dataRecord["candidates"] as Array<Record<string, unknown>> | undefined;
    const candidate = candidates?.[0];
    if (!candidate) {
      throw new Error("No candidate in response");
    }

    const candidateContent = candidate["content"] as Record<string, unknown> | undefined;
    const parts = (candidateContent?.["parts"] as unknown[]) || [];
    const content = this.extractContent(parts);
    // Extract tool calls for potential future use
    const toolCalls = this.extractToolCalls(parts);

    // Extract thinking content (for Gemini thinking models)
    const thinkingContent = this.extractThinkingContent(parts);

    // Extract thoughtsTokenCount for thinking models
    const usageMetadata = dataRecord["usageMetadata"] as Record<string, number> | undefined;
    const thoughtsTokenCount = usageMetadata?.["thoughtsTokenCount"] || 0;

    return {
      id: (dataRecord["id"] as string) || "unknown",
      model: config.profile.model,
      content,
      message: {
        role: "assistant",
        content,
        toolCalls: toolCalls.length > 0 ? toolCalls : undefined,
      },
      toolCalls: toolCalls.length > 0 ? toolCalls : undefined,
      usage: usageMetadata
        ? {
            promptTokens: usageMetadata["promptTokenCount"] || 0,
            completionTokens: usageMetadata["candidatesTokenCount"] || 0,
            totalTokens: usageMetadata["totalTokenCount"] || 0,
            // Add the `reasoningTokens` field to map Gemini's `thoughtsTokenCount`
            reasoningTokens: thoughtsTokenCount > 0 ? thoughtsTokenCount : undefined,
          }
        : undefined,
      finishReason: (candidate["finishReason"] as string) || "stop",
      duration: 0,
      metadata: {
        finishReason: candidate["finishReason"],
        safetyRatings: candidate["safetyRatings"],
      },
      reasoningContent: thinkingContent,
      reasoningTokens: thoughtsTokenCount > 0 ? thoughtsTokenCount : undefined,
    };
  }

  /**
   * Parsing stream response lines (rewritten)
   *
   * The Gemini Native API returns JSON directly, without the `data:` prefix.
   */
  override parseStreamLine(line: string, config: FormatterConfig): ParseStreamChunkResult {
    // Skip blank lines
    if (!line) {
      return { chunk: { done: false }, valid: false };
    }

    // The Gemini Native API returns JSON directly, without the "data:" prefix.
    try {
      const data = JSON.parse(line);
      return this.parseStreamChunk(data);
    } catch {
      // Skip invalid JSON
      return { chunk: { done: false }, valid: false };
    }
  }

  parseStreamChunk(data: unknown): ParseStreamChunkResult {
    const dataRecord = data as Record<string, unknown>;
    const candidates = dataRecord["candidates"] as Array<Record<string, unknown>> | undefined;
    const candidate = candidates?.[0];
    if (!candidate) {
      return { chunk: { done: false }, valid: false };
    }

    const candidateContent = candidate["content"] as Record<string, unknown> | undefined;
    const parts = (candidateContent?.["parts"] as unknown[]) || [];
    const content = this.extractContent(parts);
    // Extract tool calls from streaming chunks
    const toolCallsDelta = this.extractToolCalls(parts);

    // Extract thinking content delta (for Gemini thinking models)
    const thinkingDelta = this.extractThinkingDelta(parts);

    // Extract thoughtsTokenCount for thinking models
    const usageMetadata = dataRecord["usageMetadata"] as Record<string, number> | undefined;
    const thoughtsTokenCount = usageMetadata?.["thoughtsTokenCount"] || 0;

    // Handle finishReason: STOP, MAX_TOKENS, SAFETY, RECITATION, OTHER
    const finishReason = candidate["finishReason"] as string;
    const isDone =
      finishReason === "STOP" ||
      finishReason === "MAX_TOKENS" ||
      finishReason === "SAFETY" ||
      finishReason === "RECITATION" ||
      finishReason === "OTHER";

    return {
      chunk: {
        delta: content,
        done: isDone,
        usage: usageMetadata
          ? {
              promptTokens: usageMetadata["promptTokenCount"] || 0,
              completionTokens: usageMetadata["candidatesTokenCount"] || 0,
              totalTokens: usageMetadata["totalTokenCount"] || 0,
              // Add the `reasoningTokens` field to map Gemini's `thoughtsTokenCount`
              reasoningTokens: thoughtsTokenCount > 0 ? thoughtsTokenCount : undefined,
            }
          : undefined,
        finishReason,
        raw: data,
        reasoningDelta: thinkingDelta,
        toolCallsDelta: toolCallsDelta.length > 0 ? toolCallsDelta : undefined,
      },
      valid: true,
    };
  }

  convertTools(tools: ToolSchema[]): unknown {
    return convertToolsToGeminiFormat(tools);
  }

  /**
   * Convert message format
   */
  convertMessages(messages: LLMMessage[]): unknown[] {
    return messages
      .filter(msg => msg.role !== "system") // Gemini uses systemInstruction to handle system messages.
      .map(msg => {
        const converted: { role: string; parts: unknown[] } = {
          role: msg.role === "assistant" ? "model" : "user",
          parts: [],
        };

        // Process the content.
        if (typeof msg.content === "string") {
          converted.parts.push({
            text: msg.content,
          });
        } else if (Array.isArray(msg.content)) {
          converted.parts = msg.content;
        }

        // Processing the results of tool calls
        if (msg.role === "tool" && msg.toolCallId) {
          converted.role = "user";
          converted.parts = [
            {
              functionResponse: {
                name: msg.toolCallId,
                response: {
                  result:
                    typeof msg.content === "string"
                      ? msg.content
                      : JSON.parse(JSON.stringify(msg.content)),
                },
              },
            },
          ];
        }

        // Handle tool calls
        if (msg.toolCalls && msg.toolCalls.length > 0) {
          converted.parts = msg.toolCalls.map(call => ({
            functionCall: {
              name: call.function.name,
              args: JSON.parse(call.function.arguments),
            },
          }));
        }

        return converted;
      });
  }

  parseToolCalls(toolCalls: unknown[]): LLMToolCall[] {
    if (!toolCalls) return [];
    return toolCalls.map(call => ({
      id: (call as { id?: string }).id || generateId(),
      type: "function",
      function: {
        name:
          (call as { name?: string }).name ||
          (call as { functionCall?: { name?: string } }).functionCall?.name ||
          "",
        arguments:
          typeof (call as { args?: unknown }).args === "string"
            ? (call as { args: string }).args
            : JSON.stringify(
                (call as { args?: unknown }).args ||
                  (call as { functionCall?: { args?: unknown } }).functionCall?.args ||
                  {},
              ),
      },
    }));
  }

  /**
   * Construct a request body
   */
  private buildRequestBody(request: LLMRequest, config: FormatterConfig): Record<string, unknown> {
    const parameters = this.mergeParameters(config.profile.parameters, request.parameters);

    const body: Record<string, unknown> = {
      contents: [],
      generationConfig: {
        temperature: parameters["temperature"] || 0.7,
        maxOutputTokens: parameters["max_tokens"] || 4096,
        topP: parameters["top_p"] || 1.0,
        topK: parameters["top_k"] || 40,
      },
    };

    // Processing system instructions
    const { systemMessage, filteredMessages } = extractAndFilterSystemMessages(request.messages);
    if (systemMessage) {
      body["systemInstruction"] = {
        parts: [
          {
            text:
              typeof systemMessage.content === "string"
                ? systemMessage.content
                : JSON.stringify(systemMessage.content),
          },
        ],
      };
    }
    body["contents"] = this.convertMessages(filteredMessages);

    // Add tools
    if (request.tools && request.tools.length > 0) {
      body["tools"] = this.convertTools(request.tools);
    }

    // Handle thinking configuration (for Gemini thinking models)
    // Users can set parameters.thinkingConfig = { includeThoughts: true, thinkingLevel: 3 }
    // The deepMerge will handle this correctly

    return body;
  }

  /**
   * Extract the text content.
   */
  private extractContent(parts: unknown): string {
    if (!parts || !Array.isArray(parts)) {
      return "";
    }

    return parts
      .filter(part => (part as { text?: string }).text)
      .map(part => (part as { text?: string }).text ?? "")
      .join("");
  }

  /**
   * Extract tool calls
   */
  private extractToolCalls(parts: unknown): LLMToolCall[] {
    if (!parts || !Array.isArray(parts)) {
      return [];
    }

    return parts
      .filter(part => (part as { functionCall?: unknown }).functionCall)
      .map(part => ({
        id: generateId(),
        type: "function",
        function: {
          name: (part as { functionCall?: { name?: string } }).functionCall?.name ?? "",
          arguments: JSON.stringify(
            (part as { functionCall?: { args?: unknown } }).functionCall?.args || {},
          ),
        },
      }));
  }

  /**
   * Extract thought content
   *
   * Extract content of the thought type from the Gemini response
   */
  private extractThinkingContent(parts: unknown): string | undefined {
    if (!parts || !Array.isArray(parts)) {
      return undefined;
    }

    const thoughtParts = parts.filter(part => (part as { thought?: unknown }).thought);
    if (thoughtParts.length === 0) {
      return undefined;
    }

    return thoughtParts.map(part => (part as { text?: string }).text ?? "").join("\n");
  }

  /**
   * Extract incremental content of the thought type from Gemini's streaming response
   *
   *
   */
  private extractThinkingDelta(parts: unknown): string | undefined {
    if (!parts || !Array.isArray(parts)) {
      return undefined;
    }

    const thoughtParts = parts.filter(part => (part as { thought?: unknown }).thought);
    if (thoughtParts.length === 0) {
      return undefined;
    }

    return thoughtParts.map(part => (part as { text?: string }).text ?? "").join("");
  }
}

File: crates/wf-llm/src/formatters/gemini_native.rs
use super::LlmFormatter;
use crate::error::LlmResult;
use reqwest::Method;
use std::sync::atomic::{AtomicUsize, Ordering};
use wf_types::llm::{LlmProfile, LlmRequest, LlmResult as LlmResponseType, MessageStreamEvent};
use wf_types::tool::Tool;

/// Globally unique tool call indices for Gemini streams (each `functionCall`
/// part is a complete snapshot; unique indices keep separate calls apart in
/// the accumulator).
static GEMINI_CALL_INDEX: AtomicUsize = AtomicUsize::new(0);

pub struct GeminiNativeFormatter {
    base_url: String,
}

impl Default for GeminiNativeFormatter {
    fn default() -> Self {
        Self::new()
    }
}

impl GeminiNativeFormatter {
    pub fn new() -> Self {
        Self {
            base_url: "https://generativelanguage.googleapis.com/v1beta".to_string(),
        }
    }

    pub fn with_base_url(base_url: String) -> Self {
        Self { base_url }
    }

    fn convert_messages(&self, messages: &[wf_types::message::Message]) -> Vec<serde_json::Value> {
        messages
            .iter()
            .filter_map(|msg| {
                let role = match msg.role {
                    wf_types::message::MessageRole::System => return None,
                    wf_types::message::MessageRole::User => "user",
                    wf_types::message::MessageRole::Assistant => "model",
                    wf_types::message::MessageRole::Tool => "function",
                };

                let parts = match &msg.content {
                    wf_types::message::MessageContentValue::Text(text) => {
                        if text.is_empty() && msg.tool_calls.is_none() {
                            return None;
                        }
                        let mut p = Vec::new();
                        if !text.is_empty() {
                            p.push(serde_json::json!({"text": text}));
                        }
                        if let Some(ref tool_calls) = msg.tool_calls {
                            for tc in tool_calls {
                                let args: serde_json::Value = serde_json::from_str(
                                    &tc.function.arguments,
                                )
                                .unwrap_or(serde_json::Value::Object(serde_json::Map::new()));
                                p.push(serde_json::json!({
                                    "function_call": {
                                        "name": tc.function.name,
                                        "args": args,
                                    }
                                }));
                            }
                        }
                        p
                    }
                    wf_types::message::MessageContentValue::Rich(blocks) => blocks
                        .iter()
                        .filter_map(|block| match block {
                            wf_types::message::MessageContent::Text { text } => {
                                Some(serde_json::json!({"text": text}))
                            }
                            wf_types::message::MessageContent::ToolResult { tool_result } => {
                                let content_val: serde_json::Value =
                                    serde_json::from_str(&tool_result.content).unwrap_or_else(
                                        |_| serde_json::Value::String(tool_result.content.clone()),
                                    );
                                Some(serde_json::json!({
                                    "function_response": {
                                        "name": "",
                                        "response": content_val,
                                    }
                                }))
                            }
                            _ => None,
                        })
                        .collect(),
                };

                Some(serde_json::json!({"role": role, "parts": parts}))
            })
            .collect()
    }

    fn convert_generation_config(
        &self,
        request: &LlmRequest,
        profile: &LlmProfile,
    ) -> serde_json::Value {
        let merged_params =
            crate::formatter_helpers::merge_parameters(profile, &request.parameters);

        // Defaults match the deprecated formatter.
        let mut config = serde_json::json!({
            "temperature": merged_params.get("temperature").cloned().unwrap_or_else(|| serde_json::json!(0.7)),
            "maxOutputTokens": merged_params.get("max_tokens").cloned().unwrap_or_else(|| serde_json::json!(4096)),
            "topP": merged_params.get("top_p").cloned().unwrap_or_else(|| serde_json::json!(1.0)),
            "topK": merged_params.get("top_k").cloned().unwrap_or_else(|| serde_json::json!(40)),
        });

        if let Some(stop) = merged_params.get("stop") {
            let stop_seqs: Vec<String> = stop
                .as_array()
                .map(|arr| {
                    arr.iter()
                        .filter_map(|v| v.as_str().map(String::from))
                        .collect()
                })
                .unwrap_or_default();
            config["stopSequences"] = serde_json::json!(stop_seqs);
        }

        config
    }
}

impl LlmFormatter for GeminiNativeFormatter {
    fn build_request(
        &self,
        request: &LlmRequest,
        profile: &LlmProfile,
    ) -> LlmResult<reqwest::Request> {
        let url = format!(
            "{}/models/{}:generateContent?key={}",
            profile.base_url.as_deref().unwrap_or(&self.base_url),
            profile.model,
            profile.api_key.as_deref().unwrap_or("")
        );

        let body = self.build_body(request, profile)?;

        let mut req_builder = reqwest::Client::new()
            .request(Method::POST, &url)
            .header("Content-Type", "application/json")
            .json(&body);

        req_builder = super::shared::apply_auth_and_headers(req_builder, profile, "native");

        req_builder
            .build()
            .map_err(crate::error::LlmError::HttpError)
    }

    fn parse_response(&self, body: &str, request: &LlmRequest) -> LlmResult<LlmResponseType> {
        let mut result = self.parse_gemini_response(body)?;
        if super::shared::is_text_mode(request) {
            if let Some(content) = result.content.clone() {
                let calls = super::shared::parse_text_tool_calls(request, &content);
                if !calls.is_empty() {
                    result.tool_calls = Some(calls.clone());
                    result.message.tool_calls = Some(calls);
                }
            }
        }
        Ok(result)
    }

    fn parse_stream_chunk(&self, data: &str) -> LlmResult<Option<MessageStreamEvent>> {
        if data.is_empty() {
            return Ok(None);
        }

        let json: serde_json::Value = serde_json::from_str(data)?;

        if let Some(candidates) = json.get("candidates").and_then(|v| v.as_array()) {
            if let Some(candidate) = candidates.first() {
                if let Some(content) = candidate.get("content") {
                    if let Some(parts) = content.get("parts").and_then(|v| v.as_array()) {
                        for part in parts {
                            if let Some(text) = part.get("text").and_then(|v| v.as_str()) {
                                return Ok(Some(MessageStreamEvent::Text(
                                    wf_types::llm::MessageStreamText {
                                        snapshot: String::new(),
                                        text: text.to_string(),
                                    },
                                )));
                            }
                            if let Some(thought) = part.get("thought").and_then(|v| v.as_str()) {
                                return Ok(Some(MessageStreamEvent::ReasoningText(
                                    wf_types::llm::MessageStreamReasoning {
                                        snapshot: String::new(),
                                        reasoning: thought.to_string(),
                                    },
                                )));
                            }
                            if let Some(func_call) = part.get("functionCall") {
                                let name = func_call
                                    .get("name")
                                    .and_then(|v| v.as_str())
                                    .unwrap_or("")
                                    .to_string();
                                let args = func_call
                                    .get("args")
                                    .cloned()
                                    .unwrap_or(serde_json::Value::Object(serde_json::Map::new()));
                                let arguments = serde_json::to_string(&args).unwrap_or_default();
                                return Ok(Some(MessageStreamEvent::ToolCallDelta(
                                    wf_types::llm::MessageStreamToolCallDelta {
                                        index: GEMINI_CALL_INDEX.fetch_add(1, Ordering::Relaxed),
                                        id: None,
                                        name: Some(name),
                                        arguments: Some(arguments),
                                        is_snapshot: true,
                                    },
                                )));
                            }
                        }
                    }
                }
                if candidate.get("finishReason").is_some() {
                    return Ok(Some(MessageStreamEvent::End(
                        wf_types::llm::MessageStreamEnd {},
                    )));
                }
            }
        }

        Ok(None)
    }

    fn convert_tools(&self, tools: &[Tool]) -> LlmResult<Vec<serde_json::Value>> {
        let function_declarations: Vec<serde_json::Value> = tools
            .iter()
            .map(|t| {
                serde_json::json!({
                    "name": t.name,
                    "description": t.description,
                    "parameters": t.parameters,
                })
            })
            .collect();
        Ok(function_declarations)
    }

    fn parse_tool_calls(&self, result: &LlmResponseType) -> Vec<wf_types::message::LlmToolCall> {
        result.tool_calls.clone().unwrap_or_default()
    }
}

impl GeminiNativeFormatter {
    /// Build the request body (separated from the HTTP layer for testability).
    fn build_body(
        &self,
        request: &LlmRequest,
        profile: &LlmProfile,
    ) -> LlmResult<serde_json::Value> {
        let generation_config = self.convert_generation_config(request, profile);

        let use_text_mode = super::shared::is_text_mode(request);

        // System instructions are sent in the dedicated `systemInstruction`
        // field in both modes. Text mode injects the original system + tool
        // usage instructions + declarations; native mode keeps the original
        // system message.
        let (system_content, _) = crate::tool_format::extract_system_message(&request.messages);

        let history = if use_text_mode {
            super::shared::convert_history_for_text_mode(&request.messages, request)
        } else {
            request.messages.clone()
        };
        let messages = self.convert_messages(&history);

        let mut body = serde_json::json!({
            "contents": messages,
            "generationConfig": generation_config,
        });

        if use_text_mode {
            let system = super::shared::text_mode_system_content(request);
            if !system.is_empty() {
                body["systemInstruction"] = serde_json::json!({"parts": [{"text": system}]});
            }
        } else {
            if let Some(system) = system_content {
                if !system.is_empty() {
                    body["systemInstruction"] = serde_json::json!({"parts": [{"text": system}]});
                }
            }
            if let Some(tools) = &request.tools {
                let function_declarations: Vec<serde_json::Value> = tools
                    .iter()
                    .map(|t| {
                        serde_json::json!({
                            "name": t.name,
                            "description": t.description,
                            "parameters": t.parameters,
                        })
                    })
                    .collect();
                body["tools"] =
                    serde_json::json!([{"functionDeclarations": function_declarations}]);
            }
        }

        super::shared::apply_custom_body(&mut body, profile);

        Ok(body)
    }

    fn parse_gemini_response(&self, body: &str) -> LlmResult<LlmResponseType> {
        let json: serde_json::Value = serde_json::from_str(body)?;

        let candidates = json.get("candidates").and_then(|v| v.as_array());
        let first_candidate = candidates.and_then(|c| c.first());

        let mut text_content = String::new();
        let mut reasoning_content: Option<String> = None;
        let mut tool_calls = Vec::new();

        if let Some(candidate) = first_candidate {
            let content = candidate.get("content");
            if let Some(c) = content {
                let parts = c.get("parts").and_then(|v| v.as_array());
                if let Some(parts) = parts {
                    for part in parts {
                        if let Some(text) = part.get("text").and_then(|v| v.as_str()) {
                            text_content.push_str(text);
                        }
                        if let Some(thought) = part.get("thought").and_then(|v| v.as_str()) {
                            reasoning_content
                                .get_or_insert_with(String::new)
                                .push_str(thought);
                        }
                        if let Some(func_call) = part.get("functionCall") {
                            let name = func_call
                                .get("name")
                                .and_then(|v| v.as_str())
                                .unwrap_or("")
                                .to_string();
                            let args = func_call
                                .get("args")
                                .cloned()
                                .unwrap_or(serde_json::Value::Object(serde_json::Map::new()));
                            let arguments = serde_json::to_string(&args).unwrap_or_default();
                            tool_calls.push(wf_types::message::LlmToolCall {
                                id: format!("gemini_call_{}", wf_common::generate_id()),
                                r#type: "function".to_string(),
                                function: wf_types::message::LlmFunctionCall { name, arguments },
                            });
                        }
                    }
                }
            }
        }

        let usage = json
            .get("usageMetadata")
            .map(|u| wf_types::llm::TokenUsageStats {
                prompt_tokens: u
                    .get("promptTokenCount")
                    .and_then(|v| v.as_u64())
                    .unwrap_or(0) as u32,
                completion_tokens: u
                    .get("candidatesTokenCount")
                    .and_then(|v| v.as_u64())
                    .unwrap_or(0) as u32,
                total_tokens: u
                    .get("totalTokenCount")
                    .and_then(|v| v.as_u64())
                    .unwrap_or(0) as u32,
                reasoning_tokens: None,
                cache_read_tokens: u
                    .get("cachedContentTokenCount")
                    .and_then(|v| v.as_u64())
                    .map(|r| r as u32),
                cache_write_tokens: None,
                prompt_tokens_cost: None,
                completion_tokens_cost: None,
                total_cost: None,
            });

        let message = wf_types::message::Message {
            id: wf_types::Id::new(),
            role: wf_types::message::MessageRole::Assistant,
            content: wf_types::message::MessageContentValue::Text(text_content.clone()),
            timestamp: wf_common::time::now(),
            tool_call_id: None,
            tool_name: None,
            tool_calls: if tool_calls.is_empty() {
                None
            } else {
                Some(tool_calls.clone())
            },
            thinking: None,
            metadata: None,
        };

        let finish_reason = first_candidate
            .and_then(|c| c.get("finishReason"))
            .and_then(|v| v.as_str())
            .map(String::from);

        let mut metadata = std::collections::HashMap::new();
        if let Some(fr) = &finish_reason {
            metadata.insert("finish_reason".to_string(), serde_json::json!(fr));
        }
        let metadata = if metadata.is_empty() {
            None
        } else {
            Some(metadata)
        };

        Ok(LlmResponseType {
            id: Some(wf_common::generate_id()),
            model: json
                .get("modelVersion")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .to_string(),
            content: if text_content.is_empty() {
                None
            } else {
                Some(text_content)
            },
            message,
            tool_calls: if tool_calls.is_empty() {
                None
            } else {
                Some(tool_calls)
            },
            usage,
            finish_reason,
            duration: 0,
            reasoning_content: reasoning_content.clone(),
            reasoning_tokens: None,
            metadata,
            stream_stats: None,
            warnings: None,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn function_call_part_emits_snapshot_delta() {
        let chunk = r#"{"candidates":[{"content":{"parts":[{"functionCall":{"name":"get_weather","args":{"city":"Beijing"}}}]},"finishReason":"STOP"}]}"#;
        match GeminiNativeFormatter::new()
            .parse_stream_chunk(chunk)
            .expect("chunk must parse")
        {
            Some(MessageStreamEvent::ToolCallDelta(delta)) => {
                assert_eq!(delta.name.as_deref(), Some("get_weather"));
                let args: serde_json::Value =
                    serde_json::from_str(delta.arguments.as_deref().unwrap()).unwrap();
                assert_eq!(args["city"], "Beijing");
                assert!(delta.is_snapshot);
            }
            other => panic!("expected ToolCallDelta, got {:?}", other),
        }
    }

    fn profile() -> LlmProfile {
        LlmProfile {
            id: "p1".to_string(),
            name: "test".to_string(),
            provider: wf_types::llm::LlmProvider::GeminiNative,
            model: "gemini-1.5-pro".to_string(),
            api_key: Some("sk-test".to_string()),
            base_url: None,
            parameters: None,
            timeout: None,
            max_retries: None,
            retry_delay: None,
            headers: None,
            metadata: None,
            tool_call_format: None,
            auth_type: None,
            custom_headers: None,
            custom_body: None,
            custom_body_enabled: None,
            query_params: None,
            stream_options: None,
            context_window_size: None,
        }
    }

    fn msg(role: wf_types::message::MessageRole, text: &str) -> wf_types::message::Message {
        wf_types::message::Message {
            id: wf_types::Id::new(),
            role,
            content: wf_types::message::MessageContentValue::Text(text.to_string()),
            timestamp: 0,
            tool_call_id: None,
            tool_name: None,
            tool_calls: None,
            thinking: None,
            metadata: None,
        }
    }

    fn request(
        messages: Vec<wf_types::message::Message>,
        params: Option<serde_json::Value>,
    ) -> LlmRequest {
        LlmRequest {
            profile_id: "p1".to_string(),
            messages,
            parameters: params,
            tools: None,
            tool_call_format: None,
            locked_tool_call_format: None,
            violation_policy: None,
            execution_id: None,
            stream: None,
            dead_loop_detection: None,
            protocol_auto_converted: None,
        }
    }

    #[test]
    fn native_mode_sends_system_instruction() {
        let formatter = GeminiNativeFormatter::new();
        let req = request(
            vec![
                msg(wf_types::message::MessageRole::System, "You are a helper"),
                msg(wf_types::message::MessageRole::User, "Hello"),
            ],
            None,
        );
        let body = formatter.build_body(&req, &profile()).expect("must build");

        assert_eq!(
            body["systemInstruction"]["parts"][0]["text"],
            serde_json::json!("You are a helper")
        );
        let roles: Vec<&str> = body["contents"]
            .as_array()
            .unwrap()
            .iter()
            .filter_map(|m| m["role"].as_str())
            .collect();
        assert_eq!(roles, vec!["user"], "system must not appear in contents");
    }

    #[test]
    fn native_mode_sends_tools_alongside_system() {
        let formatter = GeminiNativeFormatter::new();
        let req = request(
            vec![
                msg(wf_types::message::MessageRole::System, "You are a helper"),
                msg(wf_types::message::MessageRole::User, "Hello"),
            ],
            None,
        );
        let mut req = req;
        req.tools = Some(vec![serde_json::from_value(serde_json::json!({
            "id": wf_types::Id::new(),
            "name": "get_weather",
            "description": "Get weather",
            "tool_type": "built_in",
            "parameters": {"type": "object", "properties": {}, "required": []},
            "enabled": true
        }))
        .unwrap()]);
        let body = formatter.build_body(&req, &profile()).expect("must build");

        assert_eq!(
            body["systemInstruction"]["parts"][0]["text"],
            serde_json::json!("You are a helper")
        );
        assert_eq!(
            body["tools"][0]["functionDeclarations"][0]["name"],
            serde_json::json!("get_weather")
        );
    }

    #[test]
    fn generation_config_defaults_and_overrides() {
        let formatter = GeminiNativeFormatter::new();
        let body = formatter
            .build_body(&request(vec![], None), &profile())
            .expect("must build");

        assert_eq!(
            body["generationConfig"]["temperature"],
            serde_json::json!(0.7)
        );
        assert_eq!(
            body["generationConfig"]["maxOutputTokens"],
            serde_json::json!(4096)
        );
        assert_eq!(body["generationConfig"]["topP"], serde_json::json!(1.0));
        assert_eq!(body["generationConfig"]["topK"], serde_json::json!(40));

        let req = request(
            vec![],
            Some(serde_json::json!({"temperature": 0.2, "max_tokens": 128})),
        );
        let body = formatter.build_body(&req, &profile()).expect("must build");
        assert_eq!(
            body["generationConfig"]["temperature"],
            serde_json::json!(0.2)
        );
        assert_eq!(
            body["generationConfig"]["maxOutputTokens"],
            serde_json::json!(128)
        );
    }
}
